app/extract_plugin_tar.go

```diff
@@ -0,0 +1,81 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package app
+
+import (
+	"archive/tar"
+	"compress/gzip"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/pkg/errors"
+)
+
+// extractTarGz takes in an io.Reader containing the bytes for a .tar.gz file and
+// a destination string to extract to.
+func extractTarGz(gzipStream io.Reader, dst string) error {
+	if dst == "" {
+		return errors.New("no destination path provided")
+	}
+
+	uncompressedStream, err := gzip.NewReader(gzipStream)
+	if err != nil {
+		return errors.Wrap(err, "failed to initialize gzip reader")
+	}
+	defer uncompressedStream.Close()
+
+	tarReader := tar.NewReader(uncompressedStream)
+
+	for {
+		header, err := tarReader.Next()
+		if err == io.EOF {
+			break
+		} else if err != nil {
+			return errors.Wrap(err, "failed to read next file from archive")
+		}
+
+		// Pre-emptively check type flag to avoid reporting a misleading error in
+		// trying to sanitize the header name.
+		switch header.Typeflag {
+		case tar.TypeDir:
+		case tar.TypeReg:
+		default:
+			return fmt.Errorf("unsupported type %v in %v", header.Typeflag, header.Name)
```

Comment:
I don't think we should return an error here, but rather just move on to the next header and maybe log something. Because, again, archives created with non-GNU tar and containing vendor extensions might cause the extraction to fail unnecessarily.